File: sdk/python/auto_agent_sdk/_http.py
from __future__ import annotations

from typing import Any, Optional

import httpx


class ApiError(Exception):
    def __init__(self, status_code: int, detail: Any) -> None:
        self.status_code = status_code
        self.detail = detail
        super().__init__(f"API error {status_code}: {detail}")


def _normalize_base_url(base_url: str) -> str:
    return base_url.rstrip("/")


def _auth_headers(api_key: str) -> dict[str, str]:
    return {"Authorization": f"Bearer {api_key}"}


def _raise_for_status(response: httpx.Response) -> None:
    if response.is_success:
        return
    detail: Any
    try:
        body = response.json()
        detail = body.get("detail", body)
    except Exception:
        detail = response.text
    raise ApiError(response.status_code, detail)


class HttpTransport:
    def __init__(
        self,
        *,
        base_url: str,
        api_key: str,
        transport: httpx.BaseTransport | None = None,
        timeout: float = 30.0,
    ) -> None:
        self._base_url = _normalize_base_url(base_url)
        self._api_key = api_key
        self._transport = transport
        self._timeout = timeout

    def request(
        self,
        method: str,
        path: str,
        *,
        json: Any | None = None,
        params: dict[str, Any] | None = None,
    ) -> Any:
        url = f"{self._base_url}{path}"
        with httpx.Client(
            transport=self._transport,
            timeout=self._timeout,
            headers=_auth_headers(self._api_key),
        ) as client:
            response = client.request(method, url, json=json, params=params)
        _raise_for_status(response)
        if response.status_code == 204:
            return None
        return response.json()


class AsyncHttpTransport:
    def __init__(
        self,
        *,
        base_url: str,
        api_key: str,
        transport: httpx.AsyncBaseTransport | None = None,
        timeout: float = 30.0,
    ) -> None:
        self._base_url = _normalize_base_url(base_url)
        self._api_key = api_key
        self._transport = transport
        self._timeout = timeout

    async def request(
        self,
        method: str,
        path: str,
        *,
        json: Any | None = None,
        params: dict[str, Any] | None = None,
    ) -> Any:
        url = f"{self._base_url}{path}"
        async with httpx.AsyncClient(
            transport=self._transport,
            timeout=self._timeout,
            headers=_auth_headers(self._api_key),
        ) as client:
            response = await client.request(method, url, json=json, params=params)
        _raise_for_status(response)
        if response.status_code == 204:
            return None
        return response.json()

File: sdk/typescript/src/http.ts
export class ApiError extends Error {
  readonly statusCode: number;
  readonly detail: unknown;

  constructor(statusCode: number, detail: unknown) {
    super(`API error ${statusCode}: ${formatDetail(detail)}`);
    this.name = "ApiError";
    this.statusCode = statusCode;
    this.detail = detail;
  }
}

function formatDetail(detail: unknown): string {
  if (typeof detail === "string") {
    return detail;
  }
  try {
    return JSON.stringify(detail);
  } catch {
    return String(detail);
  }
}

export function normalizeBaseUrl(baseUrl: string): string {
  return baseUrl.replace(/\/+$/, "");
}

export type FetchFn = typeof fetch;

export interface HttpTransportOptions {
  baseUrl: string;
  apiKey: string;
  timeout?: number;
  fetch?: FetchFn;
}

export class HttpTransport {
  private readonly baseUrl: string;
  private readonly apiKey: string;
  private readonly timeout: number;
  private readonly fetchFn: FetchFn;

  constructor(options: HttpTransportOptions) {
    this.baseUrl = normalizeBaseUrl(options.baseUrl);
    this.apiKey = options.apiKey;
    this.timeout = options.timeout ?? 30_000;
    this.fetchFn = options.fetch ?? fetch;
  }

  async request<T>(
    method: string,
    path: string,
    options?: {
      json?: unknown;
      params?: Record<string, string | number | boolean | undefined>;
    },
  ): Promise<T> {
    const url = new URL(`${this.baseUrl}${path}`);
    if (options?.params) {
      for (const [key, value] of Object.entries(options.params)) {
        if (value !== undefined) {
          url.searchParams.set(key, String(value));
        }
      }
    }

    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), this.timeout);

    try {
      const response = await this.fetchFn(url, {
        method,
        headers: {
          Authorization: `Bearer ${this.apiKey}`,
          ...(options?.json !== undefined
            ? { "Content-Type": "application/json" }
            : {}),
        },
        body: options?.json !== undefined ? JSON.stringify(options.json) : undefined,
        signal: controller.signal,
      });

      if (!response.ok) {
        let detail: unknown;
        try {
          const body = (await response.json()) as { detail?: unknown };
          detail = body.detail ?? body;
        } catch {
          detail = await response.text();
        }
        throw new ApiError(response.status, detail);
      }

      if (response.status === 204) {
        return undefined as T;
      }

      return (await response.json()) as T;
    } finally {
      clearTimeout(timeoutId);
    }
  }
}
